Directed staking offchain

aoikurokawa opened the pull request:
### CLI 

https://github.com/jito-foundation/stakenet/blob/3fac2314af4d3103f14f614cff53eb8695c43f1d/utils/steward-cli/README.md#directed-stake

#### Authority
- update-authority (update `directed_stake_meta_upload`, `directed_stake_whitelist`)

#### Cranks
- rebalance-directed

#### DirectedStakeWhitelist
- init-directed-stake-whitelist
- realloc-directed-stake-whitelist
- view-directed-stake-whitelist
- add-to-directed-stake-whitelist (validator, user, protocol)
- remove-from-directed-stake-whitelist
- close-directed-stake-whitelist

#### DirectedStakeMeta
- init-directed-stake-meta
- realloc-directed-stake-meta
- view-directed-stake-meta
- compute-directed-stake-meta

#### DirectedStakeTicket
- init-directed-stake-ticket
- view-directed-stake-ticket
- view-directed-stake-tickets
- update-directed-stake-ticket
- close-directed-stake-ticket

### Keeper
- Add `crank_copy_directed_stake_targets` operation to copy all targets to `DirectedStakeMeta`
- Add `directed_rebalance` operation
- Emit `directed-stake-stats` data 
    - state: steward state
    - meta-total-stake-targets: total stake targets from `DirectedStakeMeta`
    - meta-epoch-last-updated: last epoch updated `DirectedStakeMeta`
    - stake-updated-current-epoch: number of targets that staked in current epoch
    - target-updated-current-epoch: number of targets that updated by `compute-directed-stake-meta` ix
    - cluster: testnet or mainnet-beta

ebatsell commented on README.md:
```diff
@@ -44,10 +44,10 @@ On-chain Steward accounts for JitoSOL:
+- **Rust**: 1.88.0
+- **Cargo**: 1.88.0
+- **Anchor CLI**: 0.32.1
+- **Solana CLI**: 3.0.8
```
Not using solana 3.0 or anchor 0.32 yet

ebatsell commented on sdk/src/utils/debug.rs:
```diff
@@ -45,12 +45,14 @@ pub async fn debug_send_single_transaction(
+    Rebalance = 0x01 << 5,
+    PostLoopIdle = 0x01 << 6,
+    ComputeScore = 0x01 << 7,
+    RebalanceDirectedComplete = 0x01 << 0,
```
I don't think we want to change the value of the existing StateCodes here? Used by grafana to display which state we're in

ebatsell commented on sdk/src/utils/instructions.rs:
````diff
@@ -47,3 +64,162 @@ pub fn get_create_validator_history_instructions(
+///     &signer,
+///     preferences
+/// );
+/// ```
````
think comment example is unnecessary here

gzalz commented on utils/steward-cli/src/commands/init/realloc_directed_stake_whitelist.rs:
```diff
@@ -0,0 +1,174 @@
+    let data_length = directed_stake_whitelist_account_raw.data.len();
+    let whats_left = DirectedStakeWhitelist::SIZE - data_length.min(DirectedStakeWhitelist::SIZE);
+
+    let mut reallocs_left_to_run =
```
This may be off by one. When initializing and reallocing on testnet I required two realloc commands before view could properly deserialize.

aoikurokawa replied:
Addressed [d866fde](https://github.com/jito-foundation/stakenet/pull/195/commits/d866fdeba53c48840419cd0a1249f177cd35749e)

gzalz commented on utils/steward-cli/src/commands/cranks/compute_directed_stake_meta.rs:
```diff
@@ -0,0 +1,141 @@
+    pub token_mint: Pubkey,
+}
+
+/// Computes directed stake metadata by aggregating tickets and token balances.
```
I'd love to see some summary data on how many stake targets were uploaded with this command. Otherwise seems to work great on testnet.

aoikurokawa replied:
Addressed [663968d](https://github.com/jito-foundation/stakenet/pull/195/commits/663968dd2032beccac778be292facb7b8a46b3a8)

aoikurokawa commented on keepers/stakenet-keeper/src/entries/crank_copy_directed_stake_targets.rs:
```diff
@@ -0,0 +1,53 @@
+    let chunk_size = match ixs.len() {
+        0..=160 => 1,
+        _ => 8,
+    };
```
I just borrowed the logic from here, but probably 8 is fine
https://github.com/jito-foundation/stakenet/blob/e2a172c8068b2edbab3d5e1956f50779e9bc7171/keepers/stakenet-keeper/src/operations/priority_fee_commission.rs#L182-L186

aoikurokawa commented on keepers/stakenet-keeper/src/entries/crank_steward.rs:
```diff
@@ -941,6 +945,142 @@ async fn _handle_rebalance(
+            amount,
+        );
+
+        ixs_to_run.push(instruction);
```
totally, agree! that makes more safe 👍

aoikurokawa commented on sdk/src/utils/accounts.rs:
```diff
@@ -361,6 +425,158 @@ pub async fn get_validator_list_account(
+///     &program_id
+/// )
+/// .await
+/// .unwrap();
```
Yeah, make sense!

aoikurokawa commented on sdk/src/utils/instructions.rs:
```diff
@@ -47,3 +64,162 @@ pub fn get_create_validator_history_instructions(
+    let stake_pool =
+        StakePool::deserialize(&mut stake_pool_account.data.as_slice()).map_err(|e| {
+            JitoInstructionError::Custom(format!("Failed to deserialize stake pool: {e}"))
+        })?;
```
oh, we already have one, thank you!

aoikurokawa commented on utils/steward-cli/src/commands/init/realloc_directed_stake_meta.rs:
```diff
@@ -0,0 +1,172 @@
+    let data_length = directed_staking_meta_account_raw.data.len();
+    let whats_left = DirectedStakeMeta::SIZE - data_length.min(DirectedStakeMeta::SIZE);
+
+    let mut reallocs_left_to_run =
```
Addressed [d866fde](https://github.com/jito-foundation/stakenet/pull/195/commits/d866fdeba53c48840419cd0a1249f177cd35749e)

gzalz commented:
Sorry, realized something in the directed stake meta computation. Here's a scenario:

We have directed staker S who wanted to delegate to vote pubkey V1. We copy the stake meta. Then S changes their allocation towards vote pubkey V2. We copy the stake meta.

The resulting stake meta as things are written right now will yield the 10k bps worth of target lamports for both targets. When computing the stake meta, we need to init each pre-existing target in the meta with a target lamport value of 0 in memory, and then calculate targets from token holder balances.

aoikurokawa commented:
Thanks! Addressed ☝️ issue, changed the way of creating HashMap from empty map to existing meta.
and added test for above case.

https://github.com/jito-foundation/stakenet/blob/f136eb63bde99d970568e7fcb17df362e901797c/sdk/src/utils/helpers.rs#L546-L573

gzalz commented on sdk/src/utils/instructions.rs:
````diff
@@ -47,3 +63,110 @@ pub fn get_create_validator_history_instructions(
+///
+/// ```text
+/// conversion_rate_bps = (stake_pool.total_lamports * 10,000) / pool_token_supply
+/// allocation_lamports = (allocation_jitosol * conversion_rate_bps) / 10,000
````
This is something I think would be important to unit test. We want to put some protections in place WRT to this as this will be the source of truth.
